Recommendation graph

SonicDMG opened the pull request.

SonicDMG committed:
Updated DSEConfiguration to add support for graph traversals

SonicDMG committed:
Added support for DSE graph in pom.xml

SonicDMG committed:
Added graph based recommendation engine using DSL's (Domain Specific Language)
This is a pretty good size update with a set of refactors to support the new graph API, DSL's, and some recommendation engines since dse java driver version 1.4.0.  This update includes fluentAPI graph traversal examples using DSL's.  It assumes the graph already exists within your cluster and should be automatically created from the bootstrap if you are running via the KillrVideo Docker configuration.

SonicDMG committed:
Removed some changes that I made while in development

jeffreyscarpenter commented on src/main/java/killrvideo/configuration/DSEConfiguration.java:
```diff
@@ -90,19 +96,49 @@ public MappingManager cassandraNativeClusterProduction() {
+    }
+
+    @Bean
+    public MappingManager setMappingManager(DseSession session) {
```
This doesn't seem like a set method. More like a get() method.

SonicDMG replied:
Yea, looking at both setMappingManager and setKillrVideoTraversalSource I see your point.

jeffreyscarpenter commented on src/main/java/killrvideo/events/VideoAddedHandlers.java:
```diff
@@ -37,15 +37,18 @@
+    @Inject
+    DseSession dseSession;
+
+    private DseSession session;
```
Can you delete this "session" since there is an injected "dseSession" above?

SonicDMG replied:
Yes, thanks for reminding me.  Originally, there was no dseSession, it was all a regular session object.  I wasn't sure at first if there were going to be any issues making the switch so instead of switching everything over I did what you see here.  I simply forgot to go back and make it all permanent once I things looked good.

jeffreyscarpenter commented on src/main/java/killrvideo/graph/KillrVideoTraversalDsl.java:
```diff
@@ -0,0 +1,284 @@
+     * engine hackathon team for coming up with the following traversal and passing on some
+     * working code for me to start with.
+     */
+    public default GraphTraversal<S, Map<String, Object>> recommendHackathon(
```
what if we name recommendation engine implementations based on what style of recommender they are, instead of where we got them from.

SonicDMG replied:
Yes, that is coming in the next update once I add the rest of the recommenders.  I had not decided how I wanted to name them all just yet, but I wanted to get everything tidied up with at least one full working example before I moved on to the rest.

jeffreyscarpenter commented on src/main/java/killrvideo/graph/KillrVideoTraversalDsl.java:
```diff
@@ -0,0 +1,284 @@
+                        .group().by().by(__.sack().sum())
+                        // now that we have that big map of [video: score], lets order it
+                        .order(Scope.local).by(values, decr).limit(Scope.local, recommendations).select(keys).unfold()
+                        // Ok, we have our video vertices, now lets tag on the user vertex of the user who uploaded each video
```
You had me up until this point. Not sure I understand the purpose of these last few steps from the comment.

SonicDMG replied:
Ok, maybe I can explain these a little more.  Once I've completed the order() line in the traversal I have the set of recommended video vertices, however, I also need to return the user responsible for uploading each video .  To do that I need to traverse to the user without essentially losing my place from the returned videos and I need to return both the video and user vertices in one shot.  project() allows me to do this.  I am telling project() to project a video vertex and a user vertex using the following by() statements.  For each parameter I pass to project() I must have an accompanying by() statement.  The first one by() is really by(identity()) which is effectively grabbing what is currently in my traversal which in this case are the videos.  The second by() of by(__.in(EDGE_UPLOADED)) is then traversing from each of those video vertices IN over the "uploaded" edge and returning the user vertex associated with the video.  In a way it is like having multiple JOIN statements in SQL and using as("") with select to display concatenated results from multiple tables.

jeffreyscarpenter replied:
Ok, I didn't pick up on the requirement to return user data with the video.

SonicDMG replied:
I renamed the current one to recommendByUserRating() to make it more descriptive.

jeffreyscarpenter commented on src/main/java/killrvideo/graph/KillrVideoTraversalSourceDsl.java:
```diff
@@ -0,0 +1,164 @@
+package killrvideo.graph;
```
What is the difference between this file and KillrVideoTraversalDsl.java?

SonicDMG replied:
This is a really neat question which forced me to think about how to explain what is going on.  I am using a recommended DSL implementation pattern which is explained in detail here -> https://www.datastax.com/dev/blog/gremlin-dsls-in-java-with-dse-graph, but for lighter reading 
I'll give you a snippet.
<img width="1086" alt="screen shot 2017-08-29 at 2 00 28 pm" src="https://user-images.githubusercontent.com/23346205/29836197-7d5adde4-8cc2-11e7-96e6-7c4606ca0574.png">

Honestly, I could probably do a whole blog post on this topic alone.

So, if you are just looking at the KillrVideoTraversalDSL interface and the KillrVideoTraversalSourceDSL class you see here in the pull request it might not seem obvious what is going on and what the difference between them is.  In short, these are both part of a recommended pattern using a GremlinDsl annotation to generate boilerplate code for the underlying implementation.  The interface defines the step methods (e.g out(), in(), select(), etc) for the graph traversal DSL and the class spawns KillrVideoTraversalSource instances which sit atop of GraphTraversalSource and START our DSL traversals.  We are effectively slapping our KillrVideo specific language on top of the default g.V()... gremlin traversals.  So from g.V().hasLabel("video").has("videoId", "someIdHere") to killr.videos("someIdHere").  Also, because our traversals START from those defined in our class it is always assumed you are coming "through" whatever method is defined in the class when invoking those defined in the interface.

jeffreyscarpenter commented on src/main/java/killrvideo/service/SuggestedVideosService.java:
```diff
@@ -183,16 +228,187 @@ public void getSuggestedForUser(GetSuggestedForUserRequest request, StreamObserv
+             * from our recommendation engine.
+             */
+            GraphStatement gStatement = DseGraph.statementFromTraversal(killr.users(userIdString)
+                    .recommendHackathon(100, 4, 500, 10)
```
comment to explain choice/meaning of parameter values?

SonicDMG replied:
Oh haha, since you are not in the IDE it doesn't automatically tell you what those are, lol.

SonicDMG replied:
I have javadocs in place to explain these, forgot about that.  I might just point folks to the method declaration where it already explains them.

SonicDMG replied:
I went ahead and explained them anyway

jeffreyscarpenter commented on src/main/java/killrvideo/service/SuggestedVideosService.java:
```diff
@@ -183,16 +228,187 @@ public void getSuggestedForUser(GetSuggestedForUserRequest request, StreamObserv
+                LOGGER.debug("Added video vertex, uploaded, and taggedWith edges: " + graphResultSet.all());
+
+            }  else {
+                LOGGER.debug("Error handling YouTubeVideoAdded for graph: " + ex);
```
Should there be a TODO somewhere around here for better error handling?

SonicDMG replied:
Not really IMO, but I do need to update this to output as a warning.  At this point if something errors out on the graph I want to say something, but not attempting to re-execute.  I guess to your point maybe a TODO to eventually do something like that?

SonicDMG replied:
Added TODO's and changed logger output to warn

jeffreyscarpenter commented on src/main/java/killrvideo/service/SuggestedVideosService.java:
```diff
@@ -183,16 +228,187 @@ public void getSuggestedForUser(GetSuggestedForUserRequest request, StreamObserv
+                LOGGER.debug("Added user vertex: " + graphResultSet.one());
+
+            } else {
+                LOGGER.debug("Error creating user vertex: " + ex);
```
Same comment regarding TODO for error handling.

SonicDMG replied:
Same as above ;)

SonicDMG replied:
Added TODO's and changed logger output to warn

jeffreyscarpenter commented on src/main/java/killrvideo/service/SuggestedVideosService.java:
```diff
@@ -183,16 +228,187 @@ public void getSuggestedForUser(GetSuggestedForUserRequest request, StreamObserv
+                LOGGER.debug("Added rating between user and video: " + graphResultSet.one());
+
+            } else {
+                LOGGER.debug("Error Adding rating between user and video: " + ex);
```
TODO

SonicDMG replied:
Will do TODO

SonicDMG replied:
Added TODO's and changed logger output to warn

jeffreyscarpenter commented on src/main/java/killrvideo/service/VideoCatalogService.java:
```diff
@@ -240,27 +247,29 @@ public void submitYouTubeVideo(SubmitYouTubeVideoRequest request, StreamObserver
-                         * the handle() method.  The youTubeVideoAdded type triggers the handler.
+                         * eventbus.post() for youTubeVideoAdded below is located both in the
+                         * VideoAddedhandlers and SuggestedVideos Service classes within the handle() method.
+                         * The YouTubeVideoAdded type triggers the handler.  The call in SuggestedVideos is
```
I'm a purist I guess but having comments referencing who the subscriber of an event is seems to violate some principle of separation.

SonicDMG replied:
I am doing this for those that might not be familiar with what is going on there.  I'm trying to save folks from having to look all over the place figuring out what is going on.

SonicDMG committed:
Made improvements per Jeff's review of my pull request here -> https://github.com/KillrVideo/killrvideo-java/pull/19
All integration test PASSED
Changed "set" methods to "get" methods for getMappingManager and getKillrVideoTraversalSource
Removed all references to session and replaced with the injected dseSession
Renamed graph DSL method recommendHackathon to recommendByUserRating to be more descriptive
Added more comments in various areas to help with explanations
Added TODO's around graph data inserts for future robustness additions and change errors to use logger warn

SonicDMG committed:
Made a small formatting change
